### PR TITLE
Dpd

### LINKED ISSUE #943: Python support for dpd

### PR DESCRIPTION
**Pull request by fweik**
Fixes  #943.

Description of changes:
 - Removed DPD, renamed INTER_DPD to DPD
 - Removed DPD_MASS stuff, can be absorbed into gamma, now
   the parameter definition  is the same as for langevin
 - Thermostat test for DPD
 - Python bindings for interaction part of DPD

Documentation is still missing.

~~Not ready for review yet, there is still some strange stuff going on.~~
Seems to be working now. ~~The interface is not yet perfect, the use can
add the interaction before enabling the thermostat, and it will already act.
We could either throw an exception if the corresponding thermostat is not
active, or leave the interactions inactive until the thermostat is switched on.
Now sure about that.~~ The interactions are now created in an inactive state
if the dpd thermostat is not turned on, and are turned on on thermostat change.

PR Checklist
------------
 - [x] Tests?
   - [x] Interface
   - [x] Core 
 - [ ] Docs?

### CODE REVIEW

File: src/python/espressomd/thermostat.pyx [hunk 1]
```diff
@@ -391,39 +382,29 @@ cdef class Thermostat(object):
...
             mpi_bcast_parameter(FIELD_NPTISO_G0)
             mpi_bcast_parameter(FIELD_NPTISO_GV)
-        
-            
-    IF DPD or INTER_DPD:
-        @AssertThermostatType(THERMO_DPD, THERMO_INTER_DPD)
-        def set_dpd(self, **kwargs):
+
+    IF DPD:
+        @AssertThermostatType(THERMO_DPD)
```
Comment (KaiSzuttor) [inline src/python/espressomd/thermostat.pyx]: We should allow multiple thermostats in the case of DPD

Comment (hmenke) [reply src/python/espressomd/thermostat.pyx]: If DPD is known to work with, say, Langevin, you can simply modify this line to
```python
@AssertThermostatType(THERMO_DPD,THERMO_LANGEVIN)
```
Please do not remove `@AssertThermostatType` altogether, as I'm pretty sure that DPD and LB do *not* play together nicely.

Comment (fweik) [reply src/python/espressomd/thermostat.pyx]: I'm not sure why there would be problems for any combination, what would be the problem with LB and DPD? Also we've used such combinations in production without issues.

**Comment by RudolfWeeber** [comment]
Regarding docs:
features.rst: Beyond the scope of hit pr. #1371
inter.rst: Here's the actual doc. Probably most of it is still valid, but no reference to py class yet.
introduction.rst: The features table needs to be updated
setup.rst: I suggest to leave a note in the thermostat doc,, explaining that dpd is replaced by a non-bonded ia

**Comment by fweik** [comment]
Just to clarify this, there still is a DPD thermostat, you have to set that, add the dpd particles and add an IA. The last step is new.

File: src/python/espressomd/thermostat.pyx (hunk 1 above)
Comment (KaiSzuttor) [reply src/python/espressomd/thermostat.pyx]: So we want to remove that line?

**Comment by KaiSzuttor** [comment]
Please add documentation for changes in the interface, otherwise this can be merged

File: src/python/espressomd/thermostat.pyx (hunk 1 above)
Comment (fweik) [reply src/python/espressomd/thermostat.pyx]: I think DPD should work with all the other thermostats.

File: src/core/global.hpp [hunk 2]
```diff
@@ -83,112 +83,98 @@ extern const Datafield fields[];
 #define FIELD_CELLGRID            1
```
Comment (hmenke) [inline src/core/global.hpp]: Is it possible to make this an `enum` to avoid manual enumeration?

File: src/python/espressomd/thermostat.pyx (hunk 1 above)
Comment (KaiSzuttor) [reply src/python/espressomd/thermostat.pyx]: That means removing the line is appropriate

File: src/core/global.hpp (hunk 2 above)
Comment (fweik) [reply src/core/global.hpp]: The global refactor is not merged in yet, will be enum after merge.

**Review by KaiSzuttor — APPROVED** [review]

**Comment by KaiSzuttor** [comment]
Can be merged?

**Comment by fweik** [comment]
No, there are still docs missing.

**Comment by fweik** [comment]
@KaiSzuttor

File: src/python/espressomd/interactions.pyx [hunk 3]
```diff
@@ -416,6 +471,77 @@ IF GAY_BERNE:
...
+                Friction coefficient of the parallel part
+            r_cut : float
+                Cutoff of the parallel part
+            trans_weight_function : float
+                The distance dependence of the orthogonal part,
+                either 0 (constant) or 1 (linear)
+            trans_gamma : float
+                Friction coefficient of the orthogonal part
+            trans_r_cut : float
+                Cutoff of the orthogonal part
```
Comment (KaiSzuttor) [inline src/python/espressomd/interactions.pyx]: newline

File: src/python/espressomd/interactions.pyx [hunk 4]
```diff
@@ -416,6 +471,77 @@ IF GAY_BERNE:
...
+
+        def is_active(self):
+            return (self._params["r_cut"] > 0) or (self._params["trans_r_cut"] > 0)
+
+        def set_params(self, **kwargs):
+            """ Set parameters for the DPD interaction.
+
+            Parameters
+            ----------
+
```
Comment (KaiSzuttor) [inline src/python/espressomd/interactions.pyx]: remove those newlines

File: src/python/espressomd/interactions.pyx [hunk 5]
```diff
@@ -416,6 +471,77 @@ IF GAY_BERNE:
...
+            weight_function : float
```
Comment (KaiSzuttor) [inline src/python/espressomd/interactions.pyx]: i'd appreciate using :obj:`float` etc.

File: src/features.def [hunk 6]
```diff
@@ -68,14 +67,7 @@ THERMOSTAT_IGNORE_NON_VIRTUAL   implies VIRTUAL_SITES_THERMOSTAT
...
 DPD
-TRANS_DPD                       implies DPD
-DPD_MASS_RED                    requires not DPD_MASS_LIN
-DPD_MASS_LIN                    requires not DPD_MASS_RED
-DPD_MASS                        equals DPD_MASS_RED or DPD_MASS_LIN
-DPD_MASS                        implies MASS, DPD
-
-/* Alternative implementation of DPD */
-INTER_DPD
+TUNABLE_SLIP
```
Comment (KaiSzuttor) [inline src/features.def:70]: Unrelated to DPD

**Comment by KaiSzuttor** [comment]
Minor changes need, otherwise looks good

File: doc/sphinx/inter.rst [hunk 7]
```diff
@@ -449,35 +449,37 @@ implemented currently, which means that the potential is discontinuous
...
+
+The interface for the Lennard-Jones interaction is implemented in 
+:class:`espressomd.interactions.HatInteraction`. The hat parameters
+can be set via::
 
-inter hat
+    system.non_bonded_inter[type1, type2].hat.set_params(**kwargs)
 
-This defines a simple force ramp between particles of the types and .
+This defines a simple force ramp between particle so of two types.
```
Comment (pkreissl) [inline doc/sphinx/inter.rst]: I don't get what the "[...] particle so of [...]" is supposed to mean.

Comment (KaiSzuttor) [reply doc/sphinx/inter.rst]: Obviously a typo: change to between particles of two types

File: testsuite/hat.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+#
+# ESPResSo is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Tests particle property setters/getters
```
Comment (pkreissl) [inline testsuite/hat.py]: Not true.

File: src/features.def (hunk 6 above)
Comment (fweik) [reply src/features.def:70]: Yes, but TUNABLE_SLIP does actually exists, and should be listed there.

**Review by KaiSzuttor — APPROVED** [review]

**Pull request merged**